### PR TITLE
Laboratory work #2, Pavel Shkunov - 21FPL2

### PR DESCRIPTION
**Pull request by pa-shk**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -236,6 +236,7 @@ def calculate_chi_values(expected: dict[str, float], observed: dict[str, int]) -
     return chi_dict
 
 
+
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:239]: remove

File: lab_2_keywords_cooccurrence/main.py [hunk 2]
```diff
@@ -3,12 +3,21 @@
...
-from typing import Optional, Sequence, Mapping
+from typing import Optional, Sequence, Mapping, Union, Type, Any
+import re
+from itertools import chain
 
 KeyPhrase = tuple[str, ...]
 KeyPhrases = Sequence[KeyPhrase]
 
 
+def is_valid(object, type):
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:14]: is valid what? Bad name

File: lab_2_keywords_cooccurrence/main.py [hunk 3]
```diff
@@ -17,7 +26,12 @@ def extract_phrases(text: str) -> Optional[Sequence[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not is_valid(text, str):
+        return None
+    striped_str = re.search(r'[\s\w].*[\s\w]', text, re.DOTALL)
+    if not striped_str:
+        return []
+    return re.split(r'\s[^\w\s]+\b|\b[^\w\s]+\s\W*', striped_str.group())
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:34]: be ready to explain what this expression does

File: lab_2_keywords_cooccurrence/main.py [hunk 4]
```diff
@@ -29,7 +43,21 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
...
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not is_valid(phrases, list) or not is_valid(stop_words, list):
+        return None
+    key_phrases = []
+    for sentence in phrases:
+        phrase = []
+        for word in sentence.lower().split():
+            if word not in stop_words:
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:52]: looks difficult... Need to think how it can be modified

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_keywords_cooccurrence/main.py [hunk 5]
```diff
@@ -29,7 +49,13 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_type_and_not_empty(phrases, list) or not check_type_and_not_empty(stop_words, list):
+        return None
+    modified_phrases = []
+    for sentence in phrases:
+        modified_sentence = ' '.join([word if word not in stop_words else '?' for word in sentence.lower().split()])
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:56]: This is a very long and difficult to read line.
1.  Save `sentence.lower().split()` to a variable
2.  Save the list with words to a variable
3. join the list

File: lab_2_keywords_cooccurrence/main.py [hunk 6]
```diff
@@ -29,7 +49,13 @@ def extract_candidate_keyword_phrases(phrases: Sequence[str], stop_words: Sequen
...
+        modified_phrases.append(modified_sentence)
+    return [tuple(phrase.split()) for phrase in extract_phrases('? '.join(modified_phrases))]
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:58]: again

File: lab_2_keywords_cooccurrence/main.py [hunk 7]
```diff
@@ -17,7 +28,16 @@ def extract_phrases(text: str) -> Optional[Sequence[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_type_and_not_empty(text, str):
+        return None
+    pattern = re.compile(r'''
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:33]: thanks, but I will ask it on face-to-face

File: lab_2_keywords_cooccurrence/main.py [hunk 8]
```diff
@@ -55,7 +84,15 @@ def calculate_word_degrees(candidate_keyword_phrases: KeyPhrases,
...
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (not check_type_and_not_empty(candidate_keyword_phrases, list)
+            or not check_type_and_not_empty(content_words, list)):
+        return None
+    word_degree = dict.fromkeys(content_words, 0)
+    for word in content_words:
+        for phrase in candidate_keyword_phrases:
+            if word in phrase:
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:93]: reverse the condition and use `continue`
Remove one ident for he next line

File: lab_2_keywords_cooccurrence/main.py [hunk 9]
```diff
@@ -123,7 +180,25 @@ def extract_candidate_keyword_phrases_with_adjoining(candidate_keyword_phrases:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_type_and_not_empty(candidate_keyword_phrases, list) or not check_type_and_not_empty(phrases, list):
+        return None
+
+    neighbours = []
+    for i in range(len(candidate_keyword_phrases) - 4):
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:187]: why -4? magic constant  is a bad practice

File: lab_2_keywords_cooccurrence/main.py [hunk 10]
```diff
@@ -123,7 +180,25 @@ def extract_candidate_keyword_phrases_with_adjoining(candidate_keyword_phrases:
...
+        if (candidate_keyword_phrases[i] == candidate_keyword_phrases[i + 3]
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:188]: again

File: lab_2_keywords_cooccurrence/main.py [hunk 11]
```diff
@@ -161,4 +253,10 @@ def load_stop_words(path: Path) -> Optional[Mapping[str, Sequence[str]]]:
     :param path: path to the file with stop word lists
     :return: a dictionary containing the language names and corresponding stop word lists
     """
-    pass
+    if not check_type_and_not_empty(path, Path):
+        return None
+    file = open(path, encoding='utf-8')
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:258]: where do you release `file` ? + file is a bad name - reserved

Comment (artyomtugaryov) [reply lab_2_keywords_cooccurrence/main.py:258]: use `with` and `path.open`

File: lab_2_keywords_cooccurrence/main.py [hunk 12]
```diff
@@ -152,7 +234,17 @@ def generate_stop_words(text: str, max_length: int) -> Optional[Sequence[str]]:
...
-    pass
+    if not check_type_and_not_empty(text, str) or not check_type_and_not_empty(max_length, int) or max_length < 0:
+        return None
+    tokens = clean_and_tokenize(text)
+    if tokens:
+        frequencies = calculate_frequencies(tokens)
+    else:
+        return None
+    scores = sorted(frequencies.values())
+    percentile = scores[round(len(scores) * 0.8) - 1]
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:245]: 0.8 to a variable

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_keywords_cooccurrence/main.py (hunk 7 above)
Comment (pa-shk) [reply lab_2_keywords_cooccurrence/main.py:33]: Ok. No problem

File: lab_2_keywords_cooccurrence/main.py [hunk 13]
```diff
@@ -123,7 +181,26 @@ def extract_candidate_keyword_phrases_with_adjoining(candidate_keyword_phrases:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_type_and_not_empty(candidate_keyword_phrases, list) or not check_type_and_not_empty(phrases, list):
+        return None
+
+    pairs = list(pairwise(candidate_keyword_phrases))
+    neighbours = []
+    distance = 3
```
Comment (artyomtugaryov) [inline lab_2_keywords_cooccurrence/main.py:189]: why distance is 3?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

**Review by artyomtugaryov — APPROVED** [review]

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**